docs/app-rails/application-security.md

```diff
@@ -0,0 +1,136 @@
+
+### 2 Sessions
+#### 2.1 What Are Sessions
+#### 2.2 Session Hijacking: 
+1. `config.force_ssl = true` is set for production environments.
+1. Provided the user with a logout button.
+#### 2.3 Session Storage
+1. Using Devise's Session Controller to manage session storage.
+1. Cookies stored client side do not contain sensitive information.
+1. Cookies time out in 15 minutes of inactivity with devise `config.timeout_in`.
+1. Cookies are encrypted client side.
+1. Devise uses BCrypt and the secret_key_base by default for secret hashing.
+1. **TODO:** Cognito, the auth service, is also managing sessions and if wanted to expire sessions after a certain time, regardless of activity, we’d have to do it there.
+#### 2.4 Rotating Encrypted and Signed Cookies Configurations
+1. **TODO:** We are not rotating secrets for hashing nor the hashing algorithm, could do with cognito.
+#### 2.5 Replay Attacks for CookieStore Sessions
+1. **TODO:** We’re not using a nonce generator that would protect against cookie replay attacks. The commented out code for this is located in `/app-rails/config/initializers/generate_security_policy.rb` This however ties users to a particular server/session as we don’t want to save nonces in the database.
+#### 2.6 Session Fixation
+#### 2.7 Session Fixation - Countermeasures
+1. Devise will automatically expire sessions on sign in and sign out with `config.expire_all_remember_me_on_sign_out = true`.
+#### 2.8 Session Expiry
+
+### 3 Cross-Site Request Forgery (CSRF)
+#### 3.1  CSRF Countermeasures
+1. GET, POST, DELETE, and rails’ resources are used appropriately in the `routes.rb` file.
+1. **TODO:** We’re not setting forgery protection in production despite passing a CSRF token to the client, therefore If I change my CSRF token manually in my browser, I’m not prevented from taking any actions (GET or POST requests) as a signed in user.
+
+### 4 Redirection and Files
+#### 4.1 Redirection
+1. link_tos do not route to user inputs, which is XSS vulnerable.
+1. redirect_to does not use params or user inputs in the routes.
+4.2 File Uploads
+1. **TODO:** App doesn’t currently handle file imports but that should be documented as it is a common need in Government Applications.
+
+### 5. User Management
+1. When using devise we don’t need to set has_secure_password on password fields.
+#### 5.1 Brute Forcing Accounts
+1. Username error is generic and does not indicate whether it was an error with the username or password.
+1. Forgot password confirms the email was sent, and not whether the username exists.
+#### 5.2 Account Hijacking
+1. Change password includes email 6 digit code.
+1. **TODO:** Require entering the password when changing email.
+#### 5.3 CAPTCHA
+1. **TODO:** Include Captcha on account creation, login, change password, and change email.
+1. **TODO:** Include honeypot fields and logic on Non logged in forms to catch bots that spam all fields (good resource: https://nedbatchelder.com/text/stopbots.html).
```

Comment:
Do you feel it's necessary to include honeypots for the template? I think the TODO here might be to include it as a recommended best practice and say something like:
```
    **NOTE:** Any forms added to the application that are accessible by unauthenticated users should refer to the [Rails Guide on Security, section 5.3](https://guides.rubyonrails.org/security.html#captchas).
```

Reply:
In the updated format I left it as an open checkbox. It seems like at minimum, these honeypot fields are hidden from users viewing the application on a screen. However, we should ensure that users with screen readers are informed not to fill in the honeypot field.